File: CS/WinForms-Dashboard-Custom-Properties/Modules/ChartScaleBreakModule.cs
using System;
using System.Collections.Generic;
using DevExpress.DashboardCommon;
using DevExpress.DashboardWin;
using DevExpress.Utils.Svg;
using DevExpress.XtraBars;
using DevExpress.XtraBars.Ribbon;
using DevExpress.XtraCharts;
using DevExpress.XtraReports.UI;

namespace WindowsFormsAppCustomProperties {
    public class ChartScaleBreakModule {
        public static readonly string PropertyName = "ScaleBreak";
        readonly DashboardDesigner designer;
        BarCheckItem barItem;

        public ChartScaleBreakModule(DashboardDesigner designer, SvgImage barImage = null) {
            this.designer = designer;
            RibbonControl ribbon = (RibbonControl)designer.MenuManager;
            RibbonPage page = ribbon.GetDashboardRibbonPage(DashboardBarItemCategory.ChartTools, DashboardRibbonPage.Design);
            RibbonPageGroup group = page.GetGroupByName("Custom Properties");
            if(group == null) {
                group = new RibbonPageGroup("Custom Properties") { Name = "Custom Properties" };
                page.Groups.Add(group);
            }
            barItem = CreateBarItem("Scale Break", barImage);
            group.ItemLinks.Add(barItem);
            barItem.ItemClick += ChangeCustomPropertyValue;
            designer.DashboardItemControlUpdated += Designer_DashboardItemControlUpdated;
            designer.DashboardItemSelected += Designer_DashboardItemSelected;
            designer.DashboardCustomPropertyChanged += Designer_DashboardCustomPropertyChanged;
            designer.CustomExport += Designer_CustomExport;
        }
        BarCheckItem CreateBarItem(string caption, SvgImage barImage) {
            BarCheckItem barItem = new BarCheckItem();
            barItem.Caption = caption;
            barItem.ImageOptions.SvgImage = barImage;
            return barItem;
        }
        void ChangeCustomPropertyValue(object sender, ItemClickEventArgs e) {
            DashboardItem dashboardItem = designer.SelectedDashboardItem;
            bool newValue = !Convert.ToBoolean(dashboardItem.CustomProperties.GetValue(PropertyName));
            string status = newValue ? "enabled" : "disabled";
            CustomPropertyHistoryItem historyItem = new CustomPropertyHistoryItem(dashboardItem, PropertyName, newValue.ToString(), $"Scale break for {dashboardItem.ComponentName} is {status}");
            designer.AddToHistory(historyItem);
        }
        void Designer_DashboardCustomPropertyChanged(object sender, CustomPropertyChangedEventArgs e) {
            if(e.Name == PropertyName)
                UpdateBarItem();
        }
        void Designer_DashboardItemSelected(object sender, DashboardItemSelectedEventArgs e) {
            UpdateBarItem();
        }
        void UpdateBarItem() {
            barItem.Checked = Convert.ToBoolean(designer.SelectedDashboardItem.CustomProperties.GetValue(PropertyName));
        }
        void Designer_CustomExport(object sender, CustomExportEventArgs e) {
            Dictionary<string, XRControl> controls = e.GetPrintableControls();
            foreach(var control in controls) {
                XRChart xrChart = control.Value as XRChart;
                if(xrChart != null) {
                    string itemComponentName = control.Key;
                    UpdateChartScaleBreaks(designer.Dashboard.Items[itemComponentName], xrChart.Diagram);
                }
            }
        }
        void Designer_DashboardItemControlUpdated(object sender, DashboardItemControlEventArgs e) {
            if(e.ChartControl != null)
                UpdateChartScaleBreaks(designer.Dashboard.Items[e.DashboardItemName], e.ChartControl.Diagram);
        }
        void UpdateChartScaleBreaks(DashboardItem dashboardItem, Diagram chartDiagram) {
            if(chartDiagram != null) {
                bool scaleBreakEnabled = Convert.ToBoolean(dashboardItem.CustomProperties.GetValue(PropertyName));
                ((XYDiagram)chartDiagram).SecondaryAxesY[0].AutoScaleBreaks.Enabled = scaleBreakEnabled;
            }
        }
    }
}

File: VB/WinForms-Dashboard-Custom-Properties/Modules/ChartScaleBreakModule.vb
Imports System
Imports System.Collections.Generic
Imports DevExpress.DashboardCommon
Imports DevExpress.DashboardWin
Imports DevExpress.Utils.Svg
Imports DevExpress.XtraBars
Imports DevExpress.XtraBars.Ribbon
Imports DevExpress.XtraCharts
Imports DevExpress.XtraReports.UI

Namespace WindowsFormsAppCustomProperties

    Public Class ChartScaleBreakModule

        Public Shared ReadOnly PropertyName As String = "ScaleBreak"

        Private ReadOnly designer As DashboardDesigner

        Private barItem As BarCheckItem

        Public Sub New(ByVal designer As DashboardDesigner, ByVal Optional barImage As SvgImage = Nothing)
            Me.designer = designer
            Dim ribbon As RibbonControl = CType(designer.MenuManager, RibbonControl)
            Dim page As RibbonPage = ribbon.GetDashboardRibbonPage(DashboardBarItemCategory.ChartTools, DashboardRibbonPage.Design)
            Dim group As RibbonPageGroup = page.GetGroupByName("Custom Properties")
            If group Is Nothing Then
                group = New RibbonPageGroup("Custom Properties") With {.Name = "Custom Properties"}
                page.Groups.Add(group)
            End If

            barItem = Me.CreateBarItem("Scale Break", barImage)
            group.ItemLinks.Add(barItem)
            Me.barItem.ItemClick += AddressOf ChangeCustomPropertyValue
            designer.DashboardItemControlUpdated += AddressOf Designer_DashboardItemControlUpdated
            designer.DashboardItemSelected += AddressOf Designer_DashboardItemSelected
            designer.DashboardCustomPropertyChanged += AddressOf Designer_DashboardCustomPropertyChanged
            designer.CustomExport += AddressOf Designer_CustomExport
        End Sub

        Private Function CreateBarItem(ByVal caption As String, ByVal barImage As SvgImage) As BarCheckItem
            Dim barItem As BarCheckItem = New BarCheckItem()
            barItem.Caption = caption
            barItem.ImageOptions.SvgImage = barImage
            Return barItem
        End Function

        Private Sub ChangeCustomPropertyValue(ByVal sender As Object, ByVal e As ItemClickEventArgs)
            Dim dashboardItem As DashboardItem = designer.SelectedDashboardItem
            Dim newValue As Boolean = Not Convert.ToBoolean(dashboardItem.CustomProperties.GetValue(PropertyName))
            Dim status As String = If(newValue, "enabled", "disabled")
            Dim historyItem As CustomPropertyHistoryItem = New CustomPropertyHistoryItem(dashboardItem, PropertyName, newValue.ToString(), $"Scale break for {dashboardItem.ComponentName} is {status}")
            designer.AddToHistory(historyItem)
        End Sub

        Private Sub Designer_DashboardCustomPropertyChanged(ByVal sender As Object, ByVal e As CustomPropertyChangedEventArgs)
            If e.Name Is PropertyName Then UpdateBarItem()
        End Sub

        Private Sub Designer_DashboardItemSelected(ByVal sender As Object, ByVal e As DashboardItemSelectedEventArgs)
            UpdateBarItem()
        End Sub

        Private Sub UpdateBarItem()
            If designer.SelectedDashboardItem IsNot Nothing Then barItem.Checked = Convert.ToBoolean(designer.SelectedDashboardItem.CustomProperties.GetValue(PropertyName))
        End Sub

        Private Sub Designer_CustomExport(ByVal sender As Object, ByVal e As CustomExportEventArgs)
            Dim controls As Dictionary(Of String, XRControl) = e.GetPrintableControls()
            For Each control In controls
                Dim xrChart As XRChart = TryCast(control.Value, XRChart)
                If xrChart IsNot Nothing Then
                    Dim itemComponentName As String = control.Key
                    UpdateChartScaleBreaks(designer.Dashboard.Items(itemComponentName), xrChart.Diagram)
                End If
            Next
        End Sub

        Private Sub Designer_DashboardItemControlUpdated(ByVal sender As Object, ByVal e As DashboardItemControlEventArgs)
            If e.ChartControl IsNot Nothing Then UpdateChartScaleBreaks(designer.Dashboard.Items(e.DashboardItemName), e.ChartControl.Diagram)
        End Sub

        Private Sub UpdateChartScaleBreaks(ByVal dashboardItem As DashboardItem, ByVal chartDiagram As Diagram)
            Dim diagram As XYDiagram = Nothing
            If CSharpImpl.__Assign(diagram, TryCast(chartDiagram, XYDiagram)) IsNot Nothing Then
                Dim scaleBreakEnabled As Boolean = Convert.ToBoolean(dashboardItem.CustomProperties.GetValue(PropertyName))
                diagram.SecondaryAxesY(0).AutoScaleBreaks.Enabled = scaleBreakEnabled
            End If
        End Sub

        Private Class CSharpImpl

            <Obsolete("Please refactor calling code to use normal Visual Basic assignment")>
            Shared Function __Assign(Of T)(ByRef target As T, value As T) As T
                target = value
                Return value
            End Function
        End Class
    End Class
End Namespace
